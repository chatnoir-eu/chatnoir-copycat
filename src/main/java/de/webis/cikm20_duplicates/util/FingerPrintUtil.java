package de.webis.cikm20_duplicates.util;

import java.io.Serializable;
import java.util.Arrays;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import de.webis.trec_ndd.trec_collections.CollectionDocument;
import de.webis.trec_ndd.util.NGramms;
import info.debatty.java.lsh.MinHash;
import lombok.experimental.UtilityClass;

/**
 * 
 * @author Maik Fr√∂be
 *
 */
@UtilityClass
public class FingerPrintUtil {
	public static interface Fingerprinter<T extends Comparable<T>> extends Serializable {
		public List<T> fingerprint(CollectionDocument doc);
		
		public double similarity(List<T> a, List<T> b);
	}

	@SuppressWarnings("serial")
	public static Fingerprinter<Integer> minHashFingerPrinting(long seed) {
		int dict_size = 1;
		int size = 12;
		MinHash minHash = new MinHash(size, dict_size, seed);
		
		return new Fingerprinter<Integer>() {
			@Override
			public List<Integer> fingerprint(CollectionDocument doc) {
				return IntStream.of(minHash.signature(docToElementSet(doc))).boxed().collect(Collectors.toList());
			}

			@Override
			public double similarity(List<Integer> a, List<Integer> b) {
				int[] aArray = a.stream().mapToInt(i -> i).toArray();
				int[] bArray = b.stream().mapToInt(i -> i).toArray();
				
				return minHash.similarity(aArray, bArray);
			}
		};
	}
	
	@SuppressWarnings("serial")
	public static Fingerprinter<Integer> simHashFingerPrinting(int bitsInSimHash, int k) {
		return new Fingerprinter<Integer>() {

			@Override
			public List<Integer> fingerprint(CollectionDocument doc) {
				return Arrays.asList(1,1,1,1);
			}

			@Override
			public double similarity(List<Integer> a, List<Integer> b) {
				return 1.0;
			}
		};
	}
	
	private static Set<Integer> docToElementSet(CollectionDocument doc) {
		return NGramms.build8Gramms(doc.getFullyCanonicalizedContent()).stream()
				.map(i -> i.getMd5Hash().hashCode())
				.collect(Collectors.toSet());
	}
}
