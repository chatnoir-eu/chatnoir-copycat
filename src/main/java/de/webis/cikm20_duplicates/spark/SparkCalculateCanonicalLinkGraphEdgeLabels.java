package de.webis.cikm20_duplicates.spark;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import org.apache.htrace.shaded.fasterxml.jackson.databind.ObjectMapper;
import org.apache.spark.HashPartitioner;
import org.apache.spark.Partitioner;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import com.google.common.collect.ImmutableList;

import de.webis.cikm20_duplicates.spark.SparkCanonicalLinkGraphExtraction.CanonicalLinkGraphEdge;
import de.webis.cikm20_duplicates.spark.eval.SparkAnalyzeCanonicalLinkGraph;
import lombok.SneakyThrows;
import scala.Tuple2;

public class SparkCalculateCanonicalLinkGraphEdgeLabels {

	private static final String DIR = "cikm2020/canonical-link-graph/";
	
	public static void main(String[] args) {
		String[] corpora = new String[] {/*"cw09" ,*/ "cw12"/*, "cc-2015-11"*/};
		
		try (JavaSparkContext context = context()) {
			for(String corpus : corpora) {
				JavaRDD<String> input = context.textFile(DIR + corpus);
				
				edgeLabels(input, new HashPartitioner(10000))
					.saveAsTextFile(DIR + corpus + "-calulated-edges");
			}
		}
	}
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/analyze-canonical-link-graph");
	
		return new JavaSparkContext(conf);
	}
	
	public static JavaRDD<String> edgeLabels(JavaRDD<String> input, Partitioner partitioner) {
		return input.mapToPair(i -> toPair(i))
			.filter(i -> i != null)
			.repartitionAndSortWithinPartitions(partitioner)
			.groupByKey()
			.flatMapToPair(i -> pairsForCalculation(i))
			.filter(i -> i != null && i._1() != null && i._2() != null && i._2()._1() != null && i._2()._2() != null)
			.repartitionAndSortWithinPartitions(partitioner)
			.map(i -> reportEdgeOrNull(i._2()._1(), i._2()._2()))
			.filter(i -> i != null);
	}

	private static Tuple2<String, CanonicalLinkGraphEdge> toPair(String src) {
		CanonicalLinkGraphEdge edge = CanonicalLinkGraphEdge.fromString(src);
		String url = edge.getCanonicalLink().toString();
		String host = SparkAnalyzeCanonicalLinkGraph.hostFromUrl(url);
		if(host == null || host.equalsIgnoreCase("ERROR-PARSING-URL")) {
			return null;
		}
		
		return new Tuple2<>(url, edge);
	}
	
	private static Iterator<Tuple2<String, Tuple2<CanonicalLinkGraphEdge, CanonicalLinkGraphEdge>>> pairsForCalculation(Tuple2<String, Iterable<CanonicalLinkGraphEdge>> bla) {
		List<CanonicalLinkGraphEdge> edges = new ArrayList<>(ImmutableList.copyOf(bla._2().iterator()));
		Collections.sort(edges, (a,b) -> a.getDoc().getId().compareTo(b.getDoc().getId()));
		
		Stream<Tuple2<Integer, Integer>> indizesToCompare = IntStream.range(0, edges.size()).mapToObj(i -> i)
				.flatMap(i -> IntStream.range(i+1, edges.size()).mapToObj(j -> new Tuple2<Integer, Integer>(i,j)));
	
		return indizesToCompare.map(i -> new Tuple2<>(
				edges.get(i._1()).getDoc().getId() +"-" +edges.get(i._2()).getDoc().getId(),
				new Tuple2<>(edges.get(i._1()), edges.get(i._2()))
		)).iterator();
	}

	@SneakyThrows
	private static String reportEdgeOrNull(CanonicalLinkGraphEdge a, CanonicalLinkGraphEdge b) {
		if(!a.getCanonicalLink().equals(b.getCanonicalLink()) || a.getDoc().getId().compareTo(b.getDoc().getId()) >= 0) {
			return null;
		}
		
		Map<String, Object> ret = new LinkedHashMap<>();
		ret.put("canonicalLink", a.getCanonicalLink());
		ret.put("firstDoc", a);
		ret.put("secondDoc", b);
		ret.put("s3score", SparkEnrichRelevanceTransferPairs.s3Score(a.getDoc(), b.getDoc()));
		
		return new ObjectMapper().writeValueAsString(ret);
	}
}
