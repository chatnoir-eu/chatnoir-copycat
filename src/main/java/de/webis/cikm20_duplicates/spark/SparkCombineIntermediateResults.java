package de.webis.cikm20_duplicates.spark;

import java.util.Map;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import com.fasterxml.jackson.databind.ObjectMapper;

import lombok.SneakyThrows;

public class SparkCombineIntermediateResults {
	
	public static void main(String[] args) {
		try (JavaSparkContext context = context()) {
			JavaRDD<String> input = context.textFile("cikm2020/near-duplicate-tasks-cw09-cw12");
			
			combineIntermediateResults(input, 5000)
				.saveAsTextFile("cikm2020/results/test-01");
		}
	}

	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/candidates");

		return new JavaSparkContext(conf);
	}
	
	public static JavaRDD<String> combineIntermediateResults(JavaRDD<String> inputRDD, int partitions) {
		return inputRDD
				.map(i -> toCsv(i))
				.distinct()
				.sortBy(i -> i, Boolean.TRUE, partitions);
	}

	@SneakyThrows
	@SuppressWarnings("unchecked")
	private static String toCsv(String json) {
		Map<String, Object> ret = new ObjectMapper().readValue(json, Map.class);
		String firstId = (String) ret.get("firstId");
		String secondId = (String) ret.get("secondId");
		Integer hemmingDistance = (Integer) ret.get("hemmingDistance");
		
		if(firstId == null|| secondId == null || hemmingDistance == null) {
			throw new IllegalArgumentException("");
		}
		
		return firstId + "," + secondId + "," + hemmingDistance;
	}
}
