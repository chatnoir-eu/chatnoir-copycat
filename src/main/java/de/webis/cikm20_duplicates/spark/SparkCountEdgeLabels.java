package de.webis.cikm20_duplicates.spark;

import java.math.BigInteger;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.common.math.BigIntegerMath;

import lombok.SneakyThrows;
import scala.Tuple2;

public class SparkCountEdgeLabels {

	public static void main(String[] args) {
		try (JavaSparkContext context = context()) {
			JavaRDD<String> exactDuplicates = context.textFile("cikm2020/exact-duplicates-simhash-cw09-cw12");
			JavaRDD<String> pairs = context.textFile("cikm2020/results/test-01");
		
			countEdgeLabels(exactDuplicates, pairs).saveAsTextFile("cikm2020/near-duplicate-graph/edge-count-cw09-cw12");
		}
	}
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/count-edges");

		return new JavaSparkContext(conf);
	}
	
	public static JavaRDD<String> countEdgeLabels(JavaRDD<String> exactDuplicates, JavaRDD<String> pairs) {
		JavaPairRDD<String, BigInteger> a = exactDuplicates.flatMapToPair(i -> flatten(i).iterator())
				.reduceByKey((i,j) -> i.add(j), 10)
				.filter(i -> !i._2.equals(BigInteger.ZERO));
		
		JavaPairRDD<String, BigInteger> b = pairs.mapToPair(i -> toTuple(i))
				.reduceByKey((i,j) -> i.add(j), 10)
				.filter(i -> !i._2.equals(BigInteger.ZERO));
		
		return a.union(b)
				.reduceByKey((i,j) -> i.add(j), 10)
				.filter(i -> !i._2.equals(BigInteger.ZERO))
				.map(i -> "{\"" + i._1() + "\":" + i._2() + "}");
	}

	private static Tuple2<String, BigInteger> toTuple(String a) {
		String[] parsed = a.split(",");
		if(parsed.length != 3) {
			throw new RuntimeException("Can not transform input to tuple: '" + a + "'");
		}
		
		String name;
		
		if((parsed[0].startsWith("clueweb09") && parsed[1].startsWith("clueweb12")) || (parsed[0].startsWith("clueweb12") && parsed[1].startsWith("clueweb09"))) {
			name = "clueweb09<->clueweb12";
		} else if (parsed[0].startsWith("clueweb09") && parsed[1].startsWith("clueweb09")) {
			name = "clueweb09<->clueweb09";
		} else if (parsed[0].startsWith("clueweb12") && parsed[1].startsWith("clueweb12")) {
			name = "clueweb12<->clueweb12";
		} else {
			throw new RuntimeException("Can not handle '" + a + "'.");
		}
		
		return new Tuple2<>(name +" (k=" + Integer.parseInt(parsed[2]) + ")", BigInteger.ONE);
	}
	
	@SneakyThrows
	@SuppressWarnings({ "unchecked", "rawtypes" })
	private static List<Tuple2<String, BigInteger>> flatten(String src) {
		Map<String, Object> s = new ObjectMapper().readValue(src, Map.class);
		List<String> docs = (List) s.get("equivalentDocuments");
		
		BigInteger all = edgesBetweenFullyConnectedGraph(docs.size());
		BigInteger clueWeb09 = edgesBetweenFullyConnectedGraph(docs.stream().filter(i -> i.startsWith("clueweb09")).count());
		BigInteger clueWeb12 = edgesBetweenFullyConnectedGraph(docs.stream().filter(i -> i.startsWith("clueweb12")).count());
		
		return Arrays.asList(
			new Tuple2<>("clueweb09<->clueweb09 (k=0)", clueWeb09),
			new Tuple2<>("clueweb12<->clueweb12 (k=0)", clueWeb12),
			new Tuple2<>("clueweb09<->clueweb12 (k=0)", all.subtract(clueWeb12).subtract(clueWeb09))
		);
	}
	
	private static BigInteger edgesBetweenFullyConnectedGraph(long count) {
		if(count <= 1) {
			return BigInteger.ZERO;
		}
		
		BigInteger nodeCount = BigIntegerMath.factorial((int) count);
		BigInteger denominator = BigInteger.valueOf(2l).multiply(BigIntegerMath.factorial(((int)count) -2));
		
		return nodeCount.divide(denominator);
	}
	
	
}
