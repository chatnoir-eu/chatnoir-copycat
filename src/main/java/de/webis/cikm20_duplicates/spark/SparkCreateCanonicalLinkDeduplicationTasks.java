package de.webis.cikm20_duplicates.spark;

import java.io.Serializable;
import java.net.URL;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import org.apache.spark.Partitioner;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.common.collect.ImmutableList;

import de.webis.cikm20_duplicates.spark.SparkCreateDeduplicationCandidates.DeduplicationUnit;
import de.webis.cikm20_duplicates.spark.SparkCreateDeduplicationCandidates.IntPartitioner;
import de.webis.cikm20_duplicates.util.ClientLocalDeduplication.DeduplicationTask;
import de.webis.cikm20_duplicates.util.SourceDocuments.DocumentWithFingerprint;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.SneakyThrows;
import scala.Tuple2;

public class SparkCreateCanonicalLinkDeduplicationTasks {
	
	public static void main(String[] args) {
		try (JavaSparkContext context = context()) {
			for(String corpus: new String[] {"cw12"/*, "cc-2015-11", "cc-2017-04"*/}) {
				JavaRDD<String> input = context.textFile(inputPath(corpus));
				
				urlDeduplicationTask(input, new IntPartitioner(50000))
					.saveAsTextFile(path(corpus) + "-near-duplicate-tasks");
			}
		}
	}
	
	public static String inputPath(String corpus) {
		if ("cw09-cw12".equals(corpus)) {
			return "cikm2020/document-fingerprints-final/cw*-jsonl.bzip2";
		} else if ("cw09-cw12-cc-2015-11".equals(corpus)) {
			return "cikm2020/document-fingerprints-final/{cw,cc-2015}*-jsonl.bzip2";
		}
		
		return "cikm2020/document-fingerprints-final/" + corpus +"-jsonl.bzip2";
	}
	
	public static String path(String corpus) {
		return "cikm2020/deduplication-final/64BitK3SimHashOneGramms-canonical-urls/" + corpus;
	}
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/canonical-link-deduplication-tasks");

		return new JavaSparkContext(conf);
	}

	public static JavaRDD<String> urlDeduplicationTask(JavaRDD<String> input, Partitioner partitioner) {
		return hashPartitionToDocument(input).groupByKey(partitioner)
				.flatMap(i -> workingPackages2(i._2()));
	}

	private static Iterator<String> workingPackages2(Iterable<DeduplicationUnit2> bla) {
		List<DeduplicationUnit2> tasks = new LinkedList<>(new HashSet<>(ImmutableList.copyOf(bla)));
		Map<URL, List<DeduplicationUnit>> ret = new LinkedHashMap<>();
		
		for(DeduplicationUnit2 t: tasks) {
			if(!ret.containsKey(t.url)) {
				ret.put(t.url, new ArrayList<>());
			}
			
			ret.get(t.url).add(new DeduplicationUnit(t.id, t.hashParts));
		}
		
		return ret.values().stream()
				.filter(i -> i.size() > 1)
				.map(i -> new DeduplicationTask(i).toString())
				.iterator();
	}

	private static JavaPairRDD<String, DeduplicationUnit2> hashPartitionToDocument(JavaRDD<String> docsWithCanonicalURL) {
		return docsWithCanonicalURL
				.flatMapToPair(doc -> extractCanonicalUrlToDocId(doc));
	}

	private static Iterator<Tuple2<String, DeduplicationUnit2>> extractCanonicalUrlToDocId(String src) {
		DocumentWithFingerprint doc = DocumentWithFingerprint.fromString(src);
		
		if(doc == null || doc.getCanonicalURL() == null ||doc.getCanonicalURL().toString().trim().isEmpty()) {
			return Collections.emptyIterator();
		}
		
		DeduplicationUnit2 dedupUnit = new DeduplicationUnit2(doc.getDocId(), doc.getCanonicalURL(), doc.getFingerprints().get("64BitK3SimHashOneGramms"));
		return Arrays.asList(new Tuple2<>(doc.getCanonicalURL().toString(), dedupUnit)).iterator();
	}
	
	private static Iterator<Tuple2<Integer, DeduplicationUnit2>> extractHashesToDocId(String src) {
		DocumentWithFingerprint doc = DocumentWithFingerprint.fromString(src);
		List<Tuple2<Integer, DeduplicationUnit2>> ret = new ArrayList<>();
		
		if(doc.getCanonicalURL() == null ||doc.getCanonicalURL().toString().trim().isEmpty()) {
			return ret.iterator();
		}
		
		DeduplicationUnit2 dedupUnit = new DeduplicationUnit2(doc.getDocId(), doc.getCanonicalURL(), doc.getFingerprints().get("64BitK3SimHashOneGramms"));
		
		for(Integer hashPart: dedupUnit.getHashParts()) {
			ret.add(new Tuple2<>(hashPart, dedupUnit));
		}
		
		return ret.iterator();
	}
	
	@Data
	@AllArgsConstructor
	@SuppressWarnings("serial")
	public static class DeduplicationUnit2 implements Serializable {
		private String id;
		private URL url;
		private ArrayList<Integer> hashParts;
		
		@Override
		@SneakyThrows
		public String toString() {
			return new ObjectMapper().writeValueAsString(this);
		}
	}
}
