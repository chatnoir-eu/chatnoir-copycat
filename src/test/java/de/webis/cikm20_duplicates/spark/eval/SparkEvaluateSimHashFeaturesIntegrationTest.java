package de.webis.cikm20_duplicates.spark.eval;

import java.util.Arrays;
import java.util.List;

import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.approvaltests.Approvals;
import org.junit.Before;
import org.junit.Test;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.holdenkarau.spark.testing.SharedJavaSparkContext;

import de.webis.cikm20_duplicates.spark.SparkCreateSourceDocumentsIntegrationTest;
import de.webis.cikm20_duplicates.spark.eval.SparkEvaluateSimHashFeatures.FeatureSetCandidate;
import de.webis.cikm20_duplicates.spark.eval.SparkEvaluateSimHashFeatures.SimHashDocumentFeatures;
import de.webis.cikm20_duplicates.util.FingerPrintUtil;
import de.webis.cikm20_duplicates.util.FingerPrintUtil.Fingerprinter;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import lombok.SneakyThrows;

public class SparkEvaluateSimHashFeaturesIntegrationTest extends SharedJavaSparkContext {
	@Test
	public void calculateAllNGrammCombinations() {
		CollectionDocument doc = new CollectionDocument();
		doc.setFullyCanonicalizedContent("a b c d e f g h i j k l m n");
		
		Approvals.verifyAsJson(SparkEvaluateSimHashFeatures.allFeatures(doc));
	}
	
	private Fingerprinter<Integer> fingerprinter;
	
	@Before
	public void setUp() {
		int k = 3;
		int bitsInSimHash = 64;
		fingerprinter = FingerPrintUtil.simHashFingerPrinting(bitsInSimHash, k);
	}
	
	@Test
	public void transformExamplePairsIntoSimHashValues() {
		JavaPairRDD<String, SimHashDocumentFeatures> hashToFeatures = SparkEvaluateSimHashFeatures.featureHashToDocToFeatures(input(), fingerprinter);
		
		Approvals.verifyAsJson(tmp(hashToFeatures.map(i -> i)));
	}
	
	@Test
	public void transformExamplePairsIntoFeatureSetCandidates() {
		JavaRDD<FeatureSetCandidate> hashToFeatures = SparkEvaluateSimHashFeatures.featureSetCandidates(input(), fingerprinter);
		
		Approvals.verifyAsJson(tmp(hashToFeatures.map(i -> i)));
	}
	
	@Test
	public void groundTruthFeatureSetCandidates() {
		JavaRDD<FeatureSetCandidate> hashToFeatures = SparkEvaluateSimHashFeatures.groundTruth(input(), 0.6);
		
		Approvals.verifyAsJson(tmp(hashToFeatures.map(i -> i)));
	}
	
	@Test
	public void groundTruthFeatureSetCandidatesWithHigherThreshold() {
		JavaRDD<FeatureSetCandidate> hashToFeatures = SparkEvaluateSimHashFeatures.groundTruth(input(), 0.95);
		
		Approvals.verifyAsJson(tmp(hashToFeatures.map(i -> i)));
	}
	
	@Test
	public void reportGroundTruthBasicsForLowThreshold() {
		JavaRDD<String> hashToFeatures = SparkEvaluateSimHashFeatures.reportFeatureSetEvaluation(input(), fingerprinter, 0.6);
		List<String> actual = SparkCreateSourceDocumentsIntegrationTest.sorted(hashToFeatures);
		
		Approvals.verifyAsJson(actual);
	}
	
	@Test
	public void reportGroundTruthBasicsForHighThreshold() {
		JavaRDD<String> hashToFeatures = SparkEvaluateSimHashFeatures.reportFeatureSetEvaluation(input(), fingerprinter, 0.95);
		List<String> actual = SparkCreateSourceDocumentsIntegrationTest.sorted(hashToFeatures);
		
		Approvals.verifyAsJson(actual);
	}
	
	private static List<String> tmp(JavaRDD<?> t) {
		return SparkCreateSourceDocumentsIntegrationTest.sorted(t.map(i -> json(i)));
	}
	
	@SneakyThrows
	private static String json(Object o) {
		return new ObjectMapper().writeValueAsString(o);
	}
	
	private JavaRDD<String> input() {
		return jsc().parallelize(Arrays.asList(
				doc(1, "a b c d e f g h i j k l m n", 2, "a b c d e f g h i j k l m n", 1f),
				doc(1, "a b c d e f g h i j k l m n", 3, "a b c d e f g h i j k l m", .9f),
				doc(1, "a b c d e f g h i j k l m n", 2, "a b c d e f g h i j k l m n", 1f),
				doc(4, "o p q r s t u v w x y z u v", 5, "o p q r s u u v w x y z u v", .7f),
				doc(1, "a b c d e f g h i j k l m n", 2, "a b c d e f g h i j k l m n", 1f)
		));
	}
	
	private static String doc(int id, String text, int id2, String text2, float s3score) {
		return "{\"firstDoc\":{\"doc\": {\"id\": \""+ id +"\", \"fullyCanonicalizedContent\": \"" + text + "\"}},"
				+ "\"secondDoc\":{\"doc\": {\"id\": \"" + id2 + "\", \"fullyCanonicalizedContent\": \""+ text2 +"\"}},"
				+ "\"s3score\": " + s3score + "}";
	}
}
