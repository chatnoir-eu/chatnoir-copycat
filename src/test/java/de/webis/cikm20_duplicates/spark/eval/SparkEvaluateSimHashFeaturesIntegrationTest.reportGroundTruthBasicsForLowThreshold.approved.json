[
  "{\"firstId\":\"1\",\"secondId\":\"2\",\"featureNames\":[\"1-3-gramms\",\"1-5-gramms\",\"1-8-gramms\",\"1-gramms\",\"3-5-gramms\",\"3-8-gramms\",\"3-gramms\",\"5-8-gramms\",\"5-gramms\",\"8-gramms\",\"S3\"]}",
  "{\"firstId\":\"1\",\"secondId\":\"3\",\"featureNames\":[\"S3\"]}",
  "{\"firstId\":\"4\",\"secondId\":\"5\",\"featureNames\":[\"1-gramms\",\"S3\"]}"
]